content/spin/kv-store.md

```diff
@@ -0,0 +1,355 @@
+
+ Let's get started with creating and deploying your first Spin application that uses key/value storage.
+
+## Prerequisites
+
+First, follow [this guide](./install.md) to install Spin v0.9.0 or later. To ensure you have the correct version, you can check with this command:
+
+<!-- @selectiveCpy -->
+
+```bash
+spin --version
+```
+
+## Use Spin SDK To Open And Persist Data In A Default Key/Value Store 
+
+First, we must open a default key/value store for our Spin application. This is a special store that every environment running Spin applications will make available for their application (which when running Spin applications locally, is built using [SQLite](https://www.sqlite.org/index.html)). Then we will complete the following actions: 
+- Check if key exists and retreive the corresponding value with the Spin SDK
+- List all the keys with the Spin SDK
+
+{{ tabs "sdk-type" }}
+
+{{ startTab "Rust"}}
+
+<!-- @selectiveCpy -->
+
+```rust
+
+use anyhow::Result;
+use http::{Method, StatusCode};
+use spin_sdk::{
+    http::{Request, Response},
+    http_component,
+    key_value::{Error, Store},
+};
+
+#[http_component]
+fn handle_request(req: Request) -> Result<Response> {
+    // Open the default key-value store
+    let store = Store::open_default()?;
+
+    let (status, body) = match req.method() {
+        &Method::POST => {
+            // Add the request (URI, body) tuple to the store
+            store.set(req.uri().path(), req.body().as_deref().unwrap_or(&[]))?;
+            (StatusCode::OK, None)
+        }
+        &Method::GET => {
+            // Get the value associated with the request URI, or return a 404 if it's not present
+            match store.get(req.uri().path()) {
+                Ok(value) => (StatusCode::OK, Some(value.into())),
+                Err(Error::NoSuchKey) => (StatusCode::NOT_FOUND, None),
+                Err(error) => return Err(error.into()),
+            }
+        }
+        &Method::DELETE => {
+            // Delete the value associated with the request URI, if present
+            store.delete(req.uri().path())?;
+            (StatusCode::OK, None)
+        }
+        &Method::HEAD => {
+            // Like GET, except do not return the value
+            match store.exists(req.uri().path()) {
+                Ok(true) => (StatusCode::OK, None),
+                Ok(false) => (StatusCode::NOT_FOUND, None),
+                Err(error) => return Err(error.into()),
+            }
+        }
+        // No other methods are currently supported
+        _ => (StatusCode::METHOD_NOT_ALLOWED, None),
+    };
+
+    Ok(http::Response::builder().status(status).body(body)?)
+}
+
+```
+
+{{ blockEnd }}
+
+{{ startTab "TypeScript"}}
+
+<!-- @selectiveCpy -->
+
+```typescript
```

Comment:
The two code samples do not match here.
I would like to consider having a canonical example application that we show for this feature (it doesn't have to be the one in either examples).

Reply:
Thanks @radu-matei 
I will be looking at this tomorrow for sure; I started writing/testing each of the examples to begin addressing having a canonical example and then found that other things need documenting. For example:

```bash
$ spin build  
// snip //
error[E0432]: unresolved import `spin_sdk::key_value`
could not find `key_value` in `spin_sdk`
```
Noticed users might have to do extra things like:
- `spin templates upgrade`, and/or
- update their config i.e. `go.mod` or `package.json` or `Cargo.toml`


<img width="617" alt="Screenshot 2023-02-23 at 4 18 21 pm" src="https://user-images.githubusercontent.com/9831342/220833557-2571b532-a458-4802-a5d1-2cbe2582402f.png">

<img width="631" alt="Screenshot 2023-02-23 at 4 20 12 pm" src="https://user-images.githubusercontent.com/9831342/220833801-ad296ee7-5d9c-4de1-acb5-bd3d7bc159de.png">

<img width="527" alt="Screenshot 2023-02-23 at 4 21 29 pm" src="https://user-images.githubusercontent.com/9831342/220834202-af92515b-94d0-4430-b6c3-55031f4e7532.png">

Thought I could merge this one by close of business today, but we have a ways to go yet; perhaps end of day tomorrow we can update the code, test and close this one out. :)

Reply:
Karthik and Rajat and I have completed these examples and provided curl requests for `GET`, `POST`, `HEAD` and `DELETE` in the document (via the latest push to this branch).